---
title: "Seurat_analysis"
author: "Stacy Hung"
date: "02/12/2019"
output: html_document
---

This script is based on vignettes from the Seurat website:
https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html
https://satijalab.org/seurat/v3.1/merge_vignette.html

```{r}
library(dplyr)
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(tidyr)
library(grid)
library(ggplot2)
library(sctransform)
library(cowplot)
library(patchwork) # required for saving R objects to and *.RDS file
library(scran)
library(batchelor)
library(harmony)
```

## Examine QC metrics

```{r}
library(tidyverse)
library(dplyr)

final.cohort <- read.table("~/Documents/sc_HL_relapse/data/final_cohort.plus_RLN.txt", sep = "\t", header = TRUE, fill = TRUE)
final.cohort <- filter(final.cohort, final.cohort$FinalCohort == "yes")

rln.metrics <- read.table("~/Documents/sc_HL_relapse/metrics/RLN_metrics.txt", sep = "\t", header = TRUE, fill = TRUE)
rln.metrics <- filter(rln.metrics, rln.metrics$SampleID %in% final.cohort$Sample)
rln.metrics$dataset <- "RLN"

relapse.metrics <- read.table("~/Documents/sc_HL_relapse/metrics/all.relapse.metrics.txt", sep = "\t", header = TRUE, fill = TRUE)
relapse.metrics <- filter(relapse.metrics, relapse.metrics$SampleID %in% final.cohort$Sample)
relapse.metrics$dataset <- "relapse"

primary.metrics <- read.table("~/Documents/sc_HL_relapse/metrics/all.primary.metrics.txt", sep = "\t", header = TRUE, fill = TRUE)
primary.metrics <- filter(primary.metrics, primary.metrics$SampleID %in% final.cohort$Sample)
primary.metrics$dataset <- "diagnostic"

# Liz's favourite metrics:
  # 1. cells recovered --> "Estimated.Number.of.Cells"
  # 2. mean reads per cell --> "Mean.Reads.per.Cell" 
  # 3. median genes per cell --> "Median.Genes.per.Cell"
  # 4. % sequencing saturation --> "Sequencing.Saturation"

# Other relevant metrics:
  # "Fraction.Reads.in.Cells"
  # "Total.Genes.Detected"

keeps <- c("SampleID", "Estimated.Number.of.Cells", "Mean.Reads.per.Cell", "Median.Genes.per.Cell", "Sequencing.Saturation", "Fraction.Reads.in.Cells", "Total.Genes.Detected", "dataset")
relapse.metrics <- relapse.metrics[c(keeps)]
primary.metrics <- primary.metrics[c(keeps)]
rln.metrics <- rln.metrics[c(keeps)]

# combine datasets
all.metrics <- rbind(primary.metrics, relapse.metrics, rln.metrics)

colnames(all.metrics) <- c("SampleID", "Est. Number of Cells", "Mean Reads / Cell", "Median Genes / Cell", "Sequencing Saturation (%)", "Fraction Reads in Cells (%)", "Total Genes Detected", "Dataset")

# change sample with id 04-32704 to be in the relapse cohort (from diagnostic cohort)
all.metrics <- all.metrics %>% mutate(Dataset = replace(Dataset, SampleID=='04-32704', 'relapse'))
primary.metrics <- filter(all.metrics, all.metrics$Dataset == "diagnostic")
relapse.metrics <- filter(all.metrics, all.metrics$Dataset == "relapse")
rln.metrics <- filter(all.metrics, all.metrics$Dataset == "RLN")

# sort by dataset and sampleID
all.metrics <- all.metrics[ order(all.metrics$Dataset), ]
# all.metrics$SampleID <- factor(all.metrics$SampleID)

# convert from wide to long
metrics.long <- gather(all.metrics, metric, value, "Est. Number of Cells":"Total Genes Detected")

# sorted version
data.sorted <- metrics.long %>%
  ungroup() %>%
  arrange(metric, value) %>%
  dplyr::mutate(order = row_number())

# colors: diagnostic - royal blue; relapse - light red; RLN - light purple
p.sorted <- ggplot(data.sorted, aes(x = order, y = value, fill=Dataset)) + 
      geom_bar(stat = "identity", width = 0.8) + 
      facet_wrap( ~ metric, scales = "free") + 
      xlab("") + ylab("") + 
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 3)) +
      theme(panel.spacing = unit(1, "lines")) +
      scale_fill_manual(values=c("cornflowerblue", "lightcoral", "gray40")) +
      scale_x_continuous(
        breaks = data.sorted$order,
        labels = data.sorted$SampleID,
        expand = c(0,0)
      )

# unsorted
p <- ggplot(metrics.long, aes(x = SampleID, y = value, fill=Dataset)) + 
      geom_bar(stat = "identity", width = 0.8) + 
      facet_wrap( ~ metric, scales = "free") + 
      xlab("") + ylab("") + 
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 3)) +
      theme(panel.spacing = unit(1, "lines")) +
      scale_fill_manual(values=c("cornflowerblue", "lightcoral", "gray40"))
```

## RLN cases: read in raw data, perform initial filtering and map meta data

```{r}
## Define dataset for analysis
dataset_loc <- "~/Documents/sc_HL_relapse/cellranger-3.1.0/RLN"
ids <- rln.metrics$SampleID

# Note that 10X has an aggr function that combines multiple samples into a single count matrix, but for Seurat analysis, this is actually counterproductive since we will lose sample ids, so we will read in by sample here:
d10x.data <- sapply(ids, function(i) {
  d10x <- Read10X(file.path(dataset_loc, i, ""))
  colnames(d10x) <- paste(sapply(strsplit(colnames(d10x), split="-"), '[[', 1L), i, sep="_")
                          d10x
})

experiment.data <- do.call("cbind", d10x.data)

hl_rln.merged <- CreateSeuratObject(
  experiment.data,
  project = "RLN merged",
  min.cells = 10, # keep all genes that are expressed in at least 3 cells --> might be too lenient for scTransform
  min.features = 200, # keep all cells that are expressing at least 200 transcripts
  names.field = 2, # the sample id is in the 2nd column after applying the delimiter "_"
  names.delim = "\\_" # the delimiter
)
# total of 15,771 cells across 4 samples (average 3.9K cells / sample)

hl_rln.merged[["percent.mt"]] <- PercentageFeatureSet(hl_rln.merged, pattern = "^MT-")

# Visualization transcript information by sample
VlnPlot(hl_rln.merged, 
        features = c("nFeature_RNA", "percent.mt"),
        sort = "increasing", log = FALSE,
        ncol = 2, pt.size = 0)

hl_rln.merged <- subset(hl_rln.merged, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 10)

# 15,139 cells or ~3.8K cells / sample

# add "relapse" label (in case we want to combine it with primary HL data later)
hl_rln.merged[["case.type"]] <- "RLN"

# map batch information to orig.idents
hl_rln.merged$batch <- plyr::mapvalues(
    x = hl_rln.merged$orig.ident, 
    from = final.cohort$Sample, 
    to = final.cohort$Batch
)

saveRDS(hl_rln.merged, file = "~/Documents/sc_HL_relapse/seurat/output/LATEST/RLN-merged.rds")
```

## Relapse cases: read in raw data, perform initial filtering, and map meta data

```{r}
## Define dataset for analysis
dataset_loc <- "~/Documents/sc_HL_relapse/cellranger-3.1.0/relapse"
ids <- relapse.metrics$SampleID

# Note that 10X has an aggr function that combines multiple samples into a single count matrix, but for Seurat analysis, this is actually counterproductive since we will lose sample ids, so we will read in by sample here:
d10x.data <- sapply(ids, function(i) {
  d10x <- Read10X(file.path(dataset_loc, i, ""))
  colnames(d10x) <- paste(sapply(strsplit(colnames(d10x), split="-"), '[[', 1L), i, sep="_")
                          d10x
})

experiment.data <- do.call("cbind", d10x.data)

hl_relapse.merged <- CreateSeuratObject(
  experiment.data,
  project = "hl relapse merged",
  min.cells = 10, # keep all genes that are expressed in at least 3 cells --> might be too lenient for scTransform
  min.features = 200, # keep all cells that are expressing at least 200 transcripts
  names.field = 2, # the sample id is in the 2nd column after applying the delimiter "_"
  names.delim = "\\_" # the delimiter
)
# total of 103,522 cells across 25 samples (average ~4.1K cells)

hl_relapse.merged[["percent.mt"]] <- PercentageFeatureSet(hl_relapse.merged, pattern = "^MT-")

# Visualization transcript information by sample
VlnPlot(hl_relapse.merged, 
        features = c("nFeature_RNA", "percent.mt"),
        sort = "increasing", log = FALSE,
        ncol = 2, pt.size = 0)

hl_relapse.merged <- subset(hl_relapse.merged, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 20)

# 85,166 cells or ~3.4K cells / sample

# add "relapse" label (in case we want to combine it with primary HL data later)
hl_relapse.merged[["case.type"]] <- "relapse"

# map batch information to orig.idents
hl_relapse.merged$batch <- plyr::mapvalues(
    x = hl_relapse.merged$orig.ident, 
    from = final.cohort$Sample, 
    to = final.cohort$Batch
)

saveRDS(hl_relapse.merged, file = "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_relapse-merged.rds")
```

## Diagnostic cases: read in raw data, perform initial filtering, and map meta data

```{r}
## Define dataset for analysis
dataset_loc <- "~/Documents/sc_HL_relapse/cellranger-3.1.0/primary"
ids <- primary.metrics$SampleID

# Note that 10X has an aggr function that combines multiple samples into a single count matrix, but for Seurat analysis, this is actually counterproductive since we will lose sample ids, so we will read in by sample here:
d10x.data <- sapply(ids, function(i) {
  d10x <- Read10X(file.path(dataset_loc, i, ""))
  colnames(d10x) <- paste(sapply(strsplit(colnames(d10x), split="-"), '[[', 1L), i, sep="_")
                          d10x
})

experiment.data <- do.call("cbind", d10x.data)

hl_diag.merged <- CreateSeuratObject(
  experiment.data,
  project = "hl diag merged",
  min.cells = 10, # keep all genes that are expressed in at least 3 cells --> might be too lenient for scTransform
  min.features = 200, # keep all cells that are expressing at least 200 transcripts
  names.field = 2, # the sample id is in the 2nd column after applying the delimiter "_"
  names.delim = "\\_" # the delimiter
)
# total of 78,584 cells across 16 samples (average ~4.9K cells)

hl_diag.merged[["percent.mt"]] <- PercentageFeatureSet(hl_diag.merged, pattern = "^MT-")

# Visualization transcript information by sample
VlnPlot(hl_diag.merged, 
        features = c("nFeature_RNA", "percent.mt"),
        sort = "increasing", log = FALSE,
        ncol = 2, pt.size = 0)

# probably lowering the % MT cutoff to 10% won't make a difference (vs. keeping 20% using in relapse dataset)
hl_diag.merged <- subset (hl_diag.merged, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10)

# 73,163 cells or 4.6K cells/sample
table(hl_diag.merged$orig.ident) # get specific numbers based on sample id

# add "diagnostic" label (in case we want to combine it with primary HL data later)
hl_diag.merged[["case.type"]] <- "diagnostic"

# map batch information to orig.idents 
hl_diag.merged$batch <- plyr::mapvalues(
    x = hl_diag.merged$orig.ident, 
    from = final.cohort$Sample, 
    to = final.cohort$Batch
)

saveRDS(hl_diag.merged, file = "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_diag.merged.rds")
```

## Merging everything together and integrate

```{r}
hl.all <- merge(hl_rln.merged, 
                y = c(hl_diag.merged, hl_relapse.merged), 
                add.cell.ids = c("RLN", "diagnostic", "relapse"),
                project = "hl_lsarp_relapse_plus_diag_and_rln")

# split cases by batch
hl_all.split <- SplitObject(hl.all, split.by = "batch")

hl_all.split <- lapply(X = hl_all.split, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

saveRDS(hl_all.split, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.split.rds")

# Perform the integration itself - needs to be done on GSC
hl_all.anchors <- FindIntegrationAnchors(object.list = hl_all.split, dims = 1:20)

# saveRDS(hl_all.anchors, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.anchors.rds")

# Use anchors to integrate the datasets together - also needs to be done on the GSC
hl_all.combined <- IntegrateData(anchorset = hl_all.anchors, dims = 1:20)

hl_all.combined <- readRDS("~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined.rds")
```

## Perform integrated analysis

```{r}
# read back in from the GSC
hl_all.combined <- readRDS("~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined.rds")

# run the standard workflow for visualization and clustering
hl_all.combined <- ScaleData(hl_all.combined)
hl_all.combined <- RunPCA(hl_all.combined, npcs = 30)

# UMAP and clustering
hl_all.combined <- RunUMAP(hl_all.combined, reduction = "pca", dims = 1:20)
hl_all.combined <- FindNeighbors(hl_all.combined, reduction = "pca", dims = 1:20)
hl_all.combined <- FindClusters(hl_all.combined, resolution = 0.5)

saveRDS(hl_all.combined, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.integrated.clustered.rds")
hl_all.combined <- readRDS("~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.integrated.clustered.rds")

# Visualization
# scale_fill_manual(values=c("cornflowerblue", "lightcoral", "gray40")) +
p1 <- DimPlot(hl_all.combined, 
              reduction = "umap", 
              cols = c("relapse"="lightcoral", 
                       "diagnostic"="cornflowerblue",
                       "RLN"="gray40"),
              group.by = "case.type") # split by cohort (diag vs. relapse)
p2 <- DimPlot(hl_all.combined, reduction = "umap", label = TRUE) # split by cluster

# Visualize diagnostic and relapse cases separately
p3 <- DimPlot(hl_all.combined, reduction = "umap", split.by = "case.type")

# get list of highly-variable genes
hl.all.hvg <- hl_all.combined@assays$integrated@var.features
write.table(hl.all.hvg, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl.integrated.hgv.txt", sep = "\t", quote = FALSE)

# Examine expression of component marker genes (n=47+) to help annotate clusters
# extract markers of interest
markers.to.plot <- c("CD19", "CD27", "IGHD", "IGHG1", "IGHG2", "IGHG3", "IGHG4", "IGHGP", "IGHM", "MS4A1", # B cell
                     "CD28", "CD40LG", "ICOS", "TNFRSF18", "TNFRSF8", # Co-stimulatory molecule
                     "IFNG", "IL2", "IL4", # Cytokine / chemokine
                     "GZMA", "GZMB", "GZMK", # Effector molecule
                     "CTLA4", "HAVCR2", "LAG3", "PDCD1", "TIGIT", # Inhibitory receptor
                     "CD68", "IDO1", "IL3RA", # Macrophage
                     "CCR7", "IL7R", "LEF1", # Naive T cell
                     "NCAM1", # NK cell
                     "SDC1", #Plasma cell
                     "CLEC4C", "NRP1", # Plasmacytoid DC
                     "MKI67", # Proliferation
                     "CD3D", "CD4", "CD8B", # T cell
                     "BCL6", "CCR4", "CXCL13", "GATA3", "KLRB1", "TBX21", # T helper
                     "EOMES", "ID2", # TF
                     "FOXP3", "IKZF2", "IL2RA", # T reg
                     "ICOSLG", "CD274", "CD44", "SELL", "CD34", "CXCR5" # not part of 2000 HGV (various categories)
                     )

FeaturePlot(hl_all.combined, features = markers.to.plot, min.cutoff = "q9")
```

## Extract gene expression matrix to customize heatmap 

```{r}
library(pheatmap)
library(dplyr)

# component markers that are in list of HVG:
markers.to.plot <- c("CD19", "CD27", "IGHD", "IGHG1", "IGHG2", "IGHG3", "IGHG4", "IGHGP", "IGHM", "MS4A1", # B cell
                     "CD28", "CD40LG", "ICOS", "TNFRSF18", "TNFRSF8", # Co-stimulatory molecule
                     "IFNG", "IL2", "IL4", # Cytokine / chemokine
                     "GZMA", "GZMB", "GZMK", # Effector molecule
                     "CTLA4", "HAVCR2", "LAG3", "PDCD1", "TIGIT", # Inhibitory receptor
                     "CD68", "IDO1", "IL3RA", # Macrophage
                     "CCR7", "IL7R", "LEF1", # Naive T cell
                     "NCAM1", # NK cell
                     "SDC1", #Plasma cell
                     "CLEC4C", "NRP1", # Plasmacytoid DC
                     "MKI67", # Proliferation
                     "CD3D", "CD4", "CD8B", # T cell
                     "BCL6", "CCR4", "CXCL13", "GATA3", "KLRB1", "TBX21", # T helper
                     "EOMES", "ID2", # TF
                     "FOXP3", "IKZF2", "IL2RA" # T reg
                     )

cluster.averages <- AverageExpression(hl_all.combined, 
                                      features = markers.to.plot,
                                      return.seurat = TRUE)
hl.integrated.cluster.avgs <- GetAssayData(object = cluster.averages, slot = "scale.data")
# a 1475 (features) X 22 (clusters) matrix

# write.table(hl.integrated.cluster.avgs, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl.integrated.cluster.avgs.expr.txt", sep = "\t")
# read back in if needed

hl.integrated.components.cluster.avgs <- hl.integrated.cluster.avgs[ rownames(hl.integrated.cluster.avgs) %in% markers.to.plot , ]

# create heatmap split by component
component.mappings <- read.table("~/Documents/sc_HL_relapse/data/component_genes_with_functions_v2.txt", sep = "\t", header = TRUE, fill = TRUE)
rownames(component.mappings) <- component.mappings$Gene
components.ordered <- component.mappings[order(component.mappings$Component, component.mappings$Gene), ]
component.mappings$Gene <- NULL
component.mappings$Alt_name <- NULL

# reorder expression matrix to match component mappings
hl.integrated.components.cluster.avgs.ordered <- hl.integrated.components.cluster.avgs[ match(rownames(components.ordered), rownames(hl.integrated.components.cluster.avgs)), ]

p4 <- pheatmap(hl.integrated.components.cluster.avgs.ordered, 
         annotation_row = component.mappings, 
         cluster_rows = FALSE, 
         annotation_names_row = FALSE, 
         angle_col = 45, 
         gaps_row = c(10,15,18,21,26,29,32,33,34,36,37,40,46,48), border_color = NA)

```

## Produce single-cell level heatmaps of expression for EACH cluster

```{r}
library(gridExtra)

NUM_CLUSTERS = 24

# hl.integrated.scaled <- hl_all.combined@assays$integrated@scale.data
hl.integrated.data <- hl_all.combined@assays$integrated@data
# this is a 1559 (features) X 76607 (barcodes) matrix

# retrieve expression for each cell across the component genes of interest
hl.integrated.markers <- hl.integrated.data[ rownames(hl.integrated.data) %in% markers.to.plot , ]

# create heatmap split by component
component.mappings <- read.table("~/Documents/sc_HL_relapse/data/component_genes_with_functions_v2.txt", sep = "\t", header = TRUE, fill = TRUE)
# order by component then gene name so that the heatmap legend makes sense
rownames(component.mappings) <- component.mappings$Gene
components.ordered <- component.mappings[order(component.mappings$Component, component.mappings$Gene), ]
components.ordered$Gene <- NULL
components.ordered$Alt_name <- NULL

# reorder expression matrix to match component mappings
hl.integrated.markers.ordered <- hl.integrated.markers[ row.names(components.ordered), ]

# to create separate expr matrices for each cluster, need mappings for cell --> cluster
hl_integrated.clusters <- hl_all.combined$seurat_clusters
# convert to char
clusters.char <- levels(hl_integrated.clusters)[hl_integrated.clusters] # ordered vector of cluster ids (0 through 26)
# barcodes.cluster <- attributes(hl_integrated.clusters)$names # ordered vector of barcodes (76,607 unique entries)
clusters.df <- data.frame(barcodes.cluster, clusters.char)

# extract barcodes corresponding to clusters (an array of vectors?)
# pseudocode: 

expr.markers <- vector('list', NUM_CLUSTERS)
heatmaps <- vector('list', NUM_CLUSTERS)
hl.seurat <- vector('list', NUM_CLUSTERS)

for (c in 1:NUM_CLUSTERS) {
  # scale by cluster and component genes
  hl.seurat[[c]] <- subset(hl_all.combined, idents = c-1, )
  hl.seurat[[c]] <- subset(hl.seurat[[c]], features = markers.to.plot )
  # hl.seurat[[c]] <- ScaleData(hl.seurat[[c]])
  
  # retrieve expression for each cell across the component genes of interest
  expr.temp <- hl.seurat[[c]]@assays$integrated@scale.data
  expr.markers[[c]] <- expr.temp[ rownames(expr.temp) %in% markers.to.plot , ]
  
  # reorder expression matrix to match component mappings
  expr.markers[[c]] <- expr.markers[[c]][ row.names(components.ordered), ]

  filename <- paste("~/Documents/sc_HL_relapse/seurat/figures/LATEST/heatmaps/scaled/heatmap-scaled-cluster", 
                    c-1, ".pdf", sep = "")
  pdf(filename, width = 8, height = 8)
  heatmaps[[c]] <- pheatmap(expr.markers[[c]], 
         annotation_row = components.ordered, 
         cluster_rows = FALSE, 
         annotation_names_row = FALSE, 
         angle_col = 45, 
         gaps_row = c(10,15,18,21,26,29,32,33,34,36,37,40,46,48), 
         border_color = NA, show_colnames = FALSE, scale = "none")
  dev.off()
}

# plot all 22 heatmaps
#grid.arrange(grobs = heatmaps, ncol=3)
grid.arrange(arrangeGrob(grobs = heatmaps, ncol=3))
```

## Rename clusters based on feature plots

```{r}
library(ggplot2)
library(dplyr)
library(gridExtra)

# Tomo's latest mappings (as of July 21, 2020)
hl_all.combined <- RenameIdents(hl_all.combined,
                                `0` = "B cells",
                                `1` = "Helper T cells",
                                `2` = "B cells",
                                `3` = "Naive T cells",
                                `4` = "Treg",
                                `5` = "Helper T cells",
                                `6` = "B cells",
                                `7` = "B cells",
                                `8` = "CTL",
                                `9` = "B cells",
                                `10` = "Helper T cells",
                                `11` = "Treg",
                                `12` = "Naive T cells",
                                `13` = "Naive T cells",
                                `14` = "B cells",
                                `15` = "NK cells",
                                `16` = "B and T mix",
                                `17` = "Treg proliferation",
                                `18` = "GCB",
                                `19` = "Plasma cell",
                                `20` = "Macrophage",
                                `21` = "Plasmacytoid DCs",
                                `22` = "B cells, macrophages",
                                `23` = "CTL")
hl_all.combined$celltype <- Idents(hl_all.combined)
saveRDS(hl_all.combined, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined.annotated.rds")

# output normalized expression values for differential expression analysis:
norm.expr <- GetAssayData(object = hl_all.combined[["RNA"]], slot = "data")

# using package scrattch.io
# write_dgCMatrix_csv(norm.expr, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined.norm_expr.txt", col1_name = "gene", chunk_size = 500)
# norm.expr.df <- as.data.frame(as.matrix(norm.expr))
# write.table(norm.expr, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/hl_combined.norm_expr.txt", sep = "\t", )

# Now plot the new annotations
DimPlot(hl_all.combined, reduction = "umap", label = TRUE, pt.size = 0.2)
DimPlot(hl_all.combined, reduction = "umap", label = TRUE, pt.size = 0.2, group.by = "orig.ident") # color by sample - will show batch effects (which should be very minimal if at all)

# create a barplot to show the numbers of cells in each cluster / component type
celltype.numcells <- as.data.frame(table(hl_all.combined$celltype))
colnames(celltype.numcells) <- c("celltype", "numcells")

p <- ggplot(celltype.numcells, aes(x = celltype, y = numcells)) + 
      geom_bar(stat = "identity", width = 0.8, fill="darkblue") + 
      xlab("") + ylab("") + 
      ylim(0, 78000) +
      theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
      theme(panel.spacing = unit(1, "lines")) +
      theme(legend.title = element_blank()) +
      geom_text(aes(label=numcells), hjust=0.5, vjust=-1, size=4, color="gray47")

celltype.numcells2 <- as.data.frame(table(hl_all.combined$celltype, hl_all.combined$case.type))
colnames(celltype.numcells2) <- c("celltype", "case_type", "numcells")

p.log <- ggplot(celltype.numcells2, aes(x = celltype, y = numcells, fill = case_type)) + 
      geom_bar(position = "fill", stat = "identity", width = 0.8) + 
      scale_y_continuous(labels = scales::percent_format()) +
      xlab("") + ylab("") + 
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
      theme(panel.spacing = unit(1, "lines")) + 
      theme(legend.title = element_blank()) +
      theme(legend.position="bottom")
      
grid.arrange(p, p.log)

# create a stacked barplot to show the proportion of each cell identity by sample
sample.celltype_num <- as.data.frame(table(hl_all.combined$orig.ident, hl_all.combined$celltype))
colnames(sample.celltype_num) <- c("sample_id", "cell_identity", "num_cells")

# order by proportion of naive B cells
sample.celltype_num <- sample.celltype_num %>% 
  dplyr::group_by(sample_id) %>% dplyr::mutate(total_cells = sum(as.numeric(num_cells)))
sample.celltype_num <- as.data.frame(sample.celltype_num %>% dplyr::mutate(proportion = as.numeric(num_cells) / total_cells))
# get order of samples by focusing only on proportion of naive B cells
temp <- subset(sample.celltype_num, 
               sample.celltype_num$cell_identity == "B cells")
temp <- temp[ order(temp$cell_identity, -temp$proportion) , ]
# set the order in the full dataframe
sample.celltype_num$sample_id <- factor(sample.celltype_num$sample_id, levels = temp$sample_id)

p <- ggplot(sample.celltype_num, 
            aes(x=sample_id, y=num_cells, fill=cell_identity)) +
  geom_bar(position = "fill", stat = "identity")+
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) +
  scale_fill_manual(name = "Cell Identity",
                    values = c("B cells" = "darkslateblue",
                               "Helper T cells" = "slategray4",
                               "Treg" = "hotpink3", 
                               "B cells, macrophages" = "royalblue1", 
                               "Naive T cells" = "aquamarine3",
                               "CTL" = "olivedrab4",
                               "Treg proliferation"="firebrick3",
                               "GCB"="darkorange",
                               "B and T mix"="grey44",
                               "Proliferation"="mediumpurple1",
                               "NK cells"="bisque3",
                               "Plasmacytoid DCs"="magenta3",
                               "Macrophage"="green3",
                               "Plasma cell"="purple")) +

  xlab("") +
  ylab("")

# rename clusters to only contain letters (for older versions of DotPlot)
# hl_all.combined  <- RenameIdents(hl_all.combined,
#                                  `B cells` = "B_cells",
#                                  `Helper T cells`="Helper_T_cells",
#                                  `B cells, macrophages`="B_cell_macrophages",
#                                  `Naive T cells`="Naive_T_cells",
#                                  `Treg proliferation`="Treg_proliferation",
#                                  `B and T mix`="B_T_mix",
#                                  `NK cells`="NK_cells",
#                                  `Plasmacytoid DCs`="pDCs",
#                                  `Plasma cell`="Plasma_cells"
#                                  )

# Plot expression of component genes for both case types, side-by-side

DotPlot(
    object = hl_all.combined,
    features = rev(markers.to.plot), 
    dot.scale = 8,
    split.by = "case.type", cols = c("red", "blue", "orange"), scale = TRUE) + 
  RotatedAxis()
  #scale_colour_gradient2(low = "#FF00FF", mid = "#000000", high = "#FFFF00")
```

## Find conserved markers (to help refine cluster annotations)

```{r}
library(metap)
library(MAST)
library(dplyr)

# change default identities to cell identities (cell types)
Idents(hl_all.combined) <- hl_all.combined$celltype

## start with one cluster to test (then do a for loop)
# naiveBcell.markers <- FindConservedMarkers(hl_all.combined, ident.1 = "NaÃ¯ve B cells", grouping.var = "case.type")

# iterate through all the cell identities
cell.identities <- as.vector(unique(Idents(hl_all.combined)))
# store the results in a vector of data frames
# initialize empty vector of results
markers.results <- vector('list', length(cell.identities))

# Approach 1: find markers that are conserved between two groups (cell identity vs. all other identities)
for (i in 1:length(cell.identities)) {
  markers.results[[i]] <- FindConservedMarkers(hl_all.combined, 
                                               ident.1 = cell.identities[i],  
                                               grouping.var = "case.type")
}
# Top (up to) 3 *conserved* markers for each cluster / cell type / cell identity:
#1. Naive T cells - CCR7, IL7R
#2. Tregs - CTLA4, IL32, TNFRSF18
#3. Helper T cells - TRAC, CD2
#4. Macrophages - CLEC10A, CSTA
#5. B cells - IGHM, HLA-DRA, MS4A1, CD74
#6. CTLs - CCL5, GZMK, NKG7
#7. GCBs - RGS13, CCDC144A
#8. Plasmocytoid DCs - GZMB, PTGDS
#9. NK cells - GNLY, NKG7
#10. B and T mix - no conserved markers
#11. Plasma cells - DERL3
#12. B cells, macrophages - RSAD2, IFIT3
#13. Treg proliferation - MKI67, BIRC5

# Plot top conserved marker using FeaturePlots
top.conserved.markers <- c("DERL3", "MKI67", "BIRC5")
FeaturePlot(hl_all.combined, features = top.conserved.markers, min.cutoff = "q9")

# plot in groups
# Group 1: T cells
t.markers <- c("IL7R", "IL32", "TRAC", "CCL5")
FeaturePlot(hl_all.combined, features = t.markers, min.cutoff = "q9")

# Group 2: B cells
b.markers <- c("MS4A1", "HLA-DRA", "IGHM", "CD74")
FeaturePlot(hl_all.combined, features = b.markers, min.cutoff = "q9")

# Group 3: all other cells
other.markers <- c("CSTA", # macrophages
                   "RGS13", # GCBs
                   "PTGDS", # pDCs
                   "GNLY", # NK cells
                   "MKI67", # Treg proliferation
                   "DERL3" # plasma cells
                   )
FeaturePlot(hl_all.combined, features = other.markers, min.cutoff = "q9")

# APPROACH 2: use FindAllMarkers (finds markers that are differentially expressed in each identity group vs all others)

markers.results2 <- FindAllMarkers(hl_all.combined, assay = "RNA", test.use = "MAST")
# filter for adjusted p-value < 0.05
markers.results2 <- filter(markers.results2, markers.results2$p_val_adj < 0.05)

write.table(markers.results2, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined-all_markers.txt", sep = "\t", row.names = FALSE, quote = FALSE)

# get top 3 markers for each cluster
markers.results2.top3 <- do.call(rbind, 
                lapply(split(markers.results2, markers.results2$cluster), function(x) x[order(-x$avg_logFC)[1:3],]))
markers.results2.top1 <- do.call(rbind, 
                lapply(split(markers.results2, markers.results2$cluster), function(x) x[order(-x$avg_logFC)[1],]))

# Top markers for each cluster (based on differential expression to all other clusters)
top.conserved.markers <- markers.results2.top1$gene
FeaturePlot(hl_all.combined, features = top.conserved.markers, min.cutoff = "q9")

# Visualize conserved cell type markers:
# conserved.markers <- c(t.markers, b.markers, other.markers)
dp <- DotPlot(
    object = hl_all.combined,
    features = top.conserved.markers, 
    dot.scale = 8,
    split.by = "case.type", cols = c("orange", "blue", "red")) + 
    RotatedAxis() + xlab("") + ylab("") + coord_flip()

```

## Differential expression analysis - testing

Check out https://nbisweden.github.io/excelerate-scRNAseq/session-de/session-de-methods.html for a review of DE methods that can be applied to scRNAseq data (not just those integrated in Seurat)

!!!Important!!! Use the uncorrected raw (but normalized) expression data to perform DE

```{r}
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
library(MAST) # required for DE analysis using MAST option in Seurat FindMarkers
library(DESeq2)
library(Seurat)

# change default assay to RNA
DefaultAssay(hl_all.combined) <- "RNA"
# normalize and scale data for DE testing
hl_all.combined <- NormalizeData(hl_all.combined)
hl_all.combined <- ScaleData(hl_all.combined)

hl_all.combined$celltype.case <- paste(Idents(hl_all.combined), hl_all.combined$case.type, sep = "_")
hl_all.combined$celltype <- Idents(hl_all.combined) # keep a copy of the cluster annotations
Idents(hl_all.combined) <- hl_all.combined$case.type # make case type the main identity (for DE analysis)

## try out all the differential expression methods:
# de.methods <- c("wilcox", # NA's for p-values --> generic method
#                 "bimod", # low discriminatory power with p-value (mostly 0's)
#                 "roc", # 0 or 1 are good; 0.5 is bad classifer
#                 "t", # 9 / 54 genes have p-value > 0 (and < 0.05)
#                 "LR", #  9 / 54 genes have p-value > 0 (and < 0.05)
#                 "MAST", #  2 / 54 genes have p-value > 0 (and < 0.05) --> **only scRNAseq method
#                 "DESeq2", # doesn't work --> designed for bulk RNAseq
#                 "negbinom", # doesn't work --> generic method
#                 "poisson" # doesn't work --> generic method
# )

# initialize empty vector of results
# de.results <- vector('list', length(de.methods))

# find DE genes between all diagnostic vs. all relapse cases (not cluster-specific) for each method
# for (i in 1:length(de.methods)) {
  # de.results[[i]] <- FindMarkers(hl.both.combined, ident.1 = "relapse", ident.2 = "diagnostic", 
                          # test.use = de.methods[i],
                          # min.pct = 0.25, logfc.threshold = 0.25)
# }
```

## Add labels to distinguish between early and late relapse cases

```{r}
# map clinical information to orig.idents (*remember to map using vectors!)

# pathology
hl_all.combined$path <- plyr::mapvalues(x = hl_all.combined$orig.ident, 
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$Pathology)
)
# EBV status
hl_all.combined$ebv <- plyr::mapvalues(x = hl_all.combined$orig.ident, 
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$EBVStatus)
)
# gender
hl_all.combined$sex <- plyr::mapvalues(x = hl_all.combined$orig.ident, 
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$Sex)
)
# age
hl_all.combined$age <- plyr::mapvalues(x = hl_all.combined$orig.ident,
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$Age)
)
# Stage
hl_all.combined$stage <- plyr::mapvalues(x = hl_all.combined$orig.ident,
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$Stage)
)
# early relapse status
hl_all.combined$earlyRelapse <- plyr::mapvalues(x = hl_all.combined$orig.ident,
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$EarlyRelapse)
)
# refractory status
hl_all.combined$refractory <- plyr::mapvalues(x = hl_all.combined$orig.ident,
    from = as.vector(final.cohort$Sample), to = as.vector(final.cohort$Refractory)
)

saveRDS(hl_all.combined, "~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined.annotated.meta_labelled.rds")

# read back in as needed
hl_all.combined <- readRDS("~/Documents/sc_HL_relapse/seurat/output/LATEST/hl_all.combined.annotated.meta_labelled.rds")
```

## Peform all the DE comparisons - remember to perform DE analysis on *unintegrated* data (stored in "RNA" assay) --> also allows all 20,792 genes to be tested

## For FindMarkers function, NB: group 1 vs. group 2 indicates for avg_logFC (log fold-chage of the average expression between the two groups), a POSITIVE value indicates the gene is MORE HIGHLY EXPRESSED in the FIRST group (i.e. group 1).

## NB NB: The following DE analyses look for DE genes between cohorts (e.g. all relapse vs. all diagnostic), and are not cluster-specific -- cluster-based analyses follow later.

```{r}
de.results <- vector('list', 6)

# 1. diagnostic vs. relapse (early + late)
Idents(hl_all.combined) <- hl_all.combined$case.type
de.diag_vs_rel <- FindMarkers(hl_all.combined, assay = "RNA", 
                              ident.1 = "diagnostic", 
                              ident.2 = "relapse", 
                              test.use = "MAST",
                              min.pct = 0.25, 
                              logfc.threshold = 0.25)
de.results[[1]] <- de.diag_vs_rel
# write results to output
#write.table(de.diag_vs_rel, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/rna.diag_vs_relapse.MAST.txt", sep = "\t", quote = FALSE)

# 2. diagnostic vs. early relapse
Idents(hl_all.combined) <- hl_all.combined$earlyRelapse
de.diag_vs_earlyrel <- FindMarkers(hl_all.combined, assay = "RNA", 
                              ident.1 = "diagnostic", 
                              ident.2 = "earlyRelapse", 
                              test.use = "MAST",
                              min.pct = 0.25, 
                              logfc.threshold = 0.25)
de.results[[2]] <- de.diag_vs_earlyrel
#write.table(de.diag_vs_earlyrel, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/rna.diag_vs_early_relapse.MAST.txt", sep = "\t", quote = FALSE)

# 3. diagnostic vs. late relapse
de.diag_vs_laterel <- FindMarkers(hl_all.combined, assay = "RNA", 
                              ident.1 = "diagnostic", 
                              ident.2 = "lateRelapse", 
                              test.use = "MAST",
                              min.pct = 0.25, 
                              logfc.threshold = 0.25)
de.results[[3]] <- de.diag_vs_laterel
#write.table(de.diag_vs_laterel, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/rna.diag_vs_late_relapse.MAST.txt", sep = "\t", quote = FALSE)

# 4. early relapse vs. late relapse
de.early_vs_late <- FindMarkers(hl_all.combined, assay = "RNA", 
                              ident.1 = "earlyRelapse", 
                              ident.2 = "lateRelapse", 
                              test.use = "MAST",
                              min.pct = 0.25, 
                              logfc.threshold = 0.25)
de.results[[4]] <- de.early_vs_late
#write.table(de.early_vs_late, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/rna.early_vs_late_relapse.MAST.txt", sep = "\t", quote = FALSE)

# 5. diagnostic vs. refractory
Idents(hl_all.combined) <- hl_all.combined$refractory 
de.diag_vs_refractory <- FindMarkers(hl_all.combined, assay = "RNA", 
                              ident.1 = "diagnostic", 
                              ident.2 = "relapseRefractory", 
                              test.use = "MAST",
                              min.pct = 0.25, 
                              logfc.threshold = 0.25)
de.results[[5]] <- de.diag_vs_refractory
#write.table(de.diag_vs_refractory, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/rna.diag_vs_refractory.MAST.txt", sep = "\t", quote = FALSE)

# 6. refractory vs. relapse (early + late)
de.refractory_vs_relapse <- FindMarkers(hl_all.combined, assay = "RNA", 
                              ident.1 = "relapseRefractory", 
                              ident.2 = "relapseNotRefractory", 
                              test.use = "MAST",
                              min.pct = 0.25, 
                              logfc.threshold = 0.25)
de.results[[6]] <- de.refractory_vs_relapse
#write.table(de.refractory_vs_relapse, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/rna.refractory_vs_relapse.MAST.txt", sep = "\t", quote = FALSE)
```

## Filter out DE results

```{r}
library(dplyr)

output.files <- c("diag_vs_relapse", "diag_vs_early_relapse", "diag_vs_late_relapse", "early_vs_late_relapse", "diag_vs_refractory", "refractory_vs_relapse")
output.folder <- "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/de_analysis/"

# Remove "*orf*" "^RP" and "MT-" proteins
for (i in 1:length(de.results)) {
  file.output <- paste(output.folder, output.files[i], ".txt", sep = "")
  de.results[[i]] <- de.results[[i]][ !(grepl("(^RP)|(MT-)|(*orf*)", rownames(de.results[[i]]))), ]
  write.table(de.results[[i]], file.output, sep = "\t", quote = FALSE)
}

```

## Functions to perform DE comparisons at the cluster (cell type) level

```{r}
# function to fine DE genes between all pairs

get_de_genes <- function(comparators, de.test, cell_counts.df) {
  
  # get list of comparators that have fewer than 100 cells
  cell_counts.df <- cell_counts.df[ cell_counts.df$Var1 %in% comparators, ] # only look at comparators of interest
  low_cell_counts.comparators <- cell_counts.df[ cell_counts.df$Freq < 3, ]$Var1
  
  NUM_CLUSTERS = length(comparators) / 2
  # de_results <- vector('list', NUM_CLUSTERS)
  de_results <- list()

  for (i in 1:NUM_CLUSTERS) {
    index.grp1 <- (i*2)-1
    index.grp2 <- i*2
    
    # check that both grp1 and grp2 are NOT in the low cell counts list
    if ( !(comparators[index.grp1] %in% low_cell_counts.comparators) && 
         !(comparators[index.grp2] %in% low_cell_counts.comparators)) {
      
        # slot = "data" (normalized counts)
        tempResults <- FindMarkers(hl_all.combined, 
                                   assay = "RNA", slot = "data",
                                   ident.1 = comparators[index.grp1], 
                                   ident.2 = comparators[index.grp2], 
                                   test.use = de.test, 
                                   latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"),
                                   min.pct = 0.25, 
                                   logfc.threshold = 0)
        # make gene a column
        tempResults$gene <- rownames(tempResults)
        # add comparison description as a column
        tempResults$comparison <- paste(comparators[index.grp1], "_VS_", comparators[index.grp2], sep = "")
        rownames(tempResults) <- NULL
        # change order of columns
        cols <- c("comparison", "gene", "p_val", "avg_logFC", "pct.1", "pct.2", "p_val_adj")
        tempResults <- tempResults[ cols ]
        # sort by descending fold change
        tempResults <- tempResults[order(-tempResults$avg_logFC), ]
        # filter out non-interesting genes (i.e. *orf*, ^RP, and ^MT- genes)
        tempResults <- tempResults[ !(grepl("(^RP)|(MT-)|(*orf*)", tempResults$gene)), ]
        # remove results that have an adjusted p-value of 1
        tempResults <- filter(tempResults, tempResults$p_val_adj < 1)
        
        # !!! NB: this step will fail if tempResults is empty (e.g. all adjusted p-values are 1)
        # store the processed results in the final result array
        if (nrow(tempResults) > 0) {
          de_results[[unique(tempResults$comparison)]] <- tempResults  
        }
      }
  }
  return(de_results)
}

# function to merge results into single dataframe
merge_de_results <- function(results.vector) {
  # initialize dataframe
  merged.df <- data.frame(comparison = character(),
                          gene = character(),
                          p_val = double(),
                          avg_logFC = double(),
                          pct.1 = double(),
                          pct.2 = double(),
                          p_val_adj = double())
  
  NUM_CLUSTERS = length(results.vector)
  for (i in 1:NUM_CLUSTERS) {
    merged.df <- rbind(merged.df, results.vector[[i]])
  }
  return(merged.df)
}
```

## For sanity purposes - test out the different methods: wilcox, bimod, roc, t, negbinom, poisson, LR, MAST, DESeq2

```{r}
de.test.mast <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "MAST", assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.wilcox <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "wilcox", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.bimod <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "bimod", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.roc <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "roc", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.t<- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "t", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.negbinom <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "negbinom", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.poisson <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "poisson", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.LR <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "LR", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

de.test.DESeq2 <- FindMarkers(hl.both.combined, 
                       ident.1 = "Macrophages_diagnostic", 
                       ident.2 = "Macrophages_relapse", 
                       test.use = "DESeq2", 
                       assay = "RNA", slot = "data",
                       latent.vars = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

```

## Differential expression analyses for all comparisons, by cell type (annotated clusters) and by cluster  

```{r}
## #########################################################
## Cluster-level DE analysis: diagnostic vs. relapse
## #########################################################

# define working directory:
work.dir <- "~/Documents/sc_HL_relapse/seurat/output/LATEST/DE_analysis/"

## ------ cell-type level -------

# define comparators (for annotated cell types)
hl_all.combined$celltype.case <- paste(hl_all.combined$celltype, hl_all.combined$case.type, sep = "_")
Idents(hl_all.combined) <- hl_all.combined$celltype.case
comparators <- sort(unique(hl_all.combined$celltype.case))

# remove RLN cases
comparators <- comparators[ !grepl("_RLN", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$celltype.case))

# perform DE analysis, where odd indices = diagnostic, even indices = relapse
de.diag_vs_rel.celltypes <- get_de_genes(comparators = comparators, de.test = "MAST", cell_counts.df = cell_counts.df)
de.diag_vs_rel.celltypes.df <- merge_de_results(de.diag_vs_rel.celltypes)
write.table(de.diag_vs_rel.celltypes.df, 
            paste(work.dir, "de.diag_vs_rel.cell_types.txt", sep=""),
            sep = "\t", quote = FALSE, row.names = FALSE)

## ------ cluster level -------

hl_all.combined$cluster.case <- paste(hl_all.combined$seurat_clusters, hl_all.combined$case.type, sep = "_")
Idents(hl_all.combined) <- hl_all.combined$cluster.case
comparators <- sort(unique(hl_all.combined$cluster.case))

# remove RLN cases
comparators <- comparators[ !grepl("_RLN", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$cluster.case))

# perform DE analysis, where odd indices = diagnostic, even indices = relapse
de.diag_vs_rel.clusters <- get_de_genes(comparators = comparators, de.test = "MAST", cell_counts.df = cell_counts.df)
de.diag_vs_rel.clusters.df <- merge_de_results(de.diag_vs_rel.clusters)
write.table(de.diag_vs_rel.clusters.df, 
            paste(work.dir, "de.diag_vs_rel.clusters.txt", sep=""),
            sep = "\t", quote = FALSE, row.names = FALSE)


## #########################################################
## Cluster-level DE analysis: diagnostic vs. refractory
## #########################################################

## ------ cell-type level -------

# define comparators
hl_all.combined$celltype.refractory <- paste(hl_all.combined$celltype, hl_all.combined$refractory, sep = "_")
Idents(hl_all.combined) <- hl_all.combined$celltype.refractory 
comparators <- sort(unique(hl_all.combined$celltype.refractory))

# remove comparators associated "_relapseNotRefractory" and RLN (i.e. "_NA") cases
comparators <- comparators[ !grepl("_(relapseNotRefractory|RLN)", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$celltype.refractory))

# perform the DE analysis
de.diag_vs_refractory.celltypes <- get_de_genes(comparators = comparators, de.test = "MAST", cell_counts.df = cell_counts.df)
de.diag_vs_refractory.celltypes.df <- merge_de_results(de.diag_vs_refractory.celltypes)
write.table(de.diag_vs_refractory.celltypes.df, 
            paste(work.dir, "de.diag_vs_refractory.cell_types.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)

## ------ cluster level -------

hl_all.combined$cluster.refractory <- paste(hl_all.combined$seurat_clusters, hl_all.combined$refractory, sep = "_")
Idents(hl_all.combined) <- hl_all.combined$cluster.refractory 
comparators <- sort(unique(hl_all.combined$cluster.refractory))

# remove comparators associated "_relapseNotRefractory" and RLN (i.e. "_NA") cases
comparators <- comparators[ !grepl("_(relapseNotRefractory|RLN)", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$cluster.refractory))

# perform the DE analysis
de.diag_vs_refractory.clusters <- get_de_genes(comparators = comparators, de.test = "MAST", cell_counts.df = cell_counts.df)
de.diag_vs_refractory.clusters.df <- merge_de_results(de.diag_vs_refractory.clusters)
write.table(de.diag_vs_refractory.clusters.df, 
            paste(work.dir, "de.diag_vs_refractory.clusters.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)


## #########################################################
## Cluster-level DE analysis: diagnostic vs. early relapse, diagnostic vs. late relapse
## #########################################################

## ------ cell type level -------

# define comparators
hl_all.combined$celltype.earlyrelapse <- paste(hl_all.combined$celltype, hl_all.combined$earlyRelapse, sep = "_")
Idents(hl_all.combined) <- hl_all.combined$celltype.earlyrelapse 
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$celltype.earlyrelapse))

# diagnostic vs. early relapse
comparators <- sort(unique(hl_all.combined$celltype.earlyrelapse))
# remove the "_lateRelapse" and RLN cases
comparators <- comparators[ !grepl("_(lateRelapse|RLN)", comparators)]

# perform the DE analysis
de.diag_vs_early_relapse.celltypes <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.diag_vs_early_relapse.celltypes.df <- merge_de_results(de.diag_vs_early_relapse.celltypes)
write.table(de.diag_vs_early_relapse.celltypes.df, 
            paste(work.dir, "de.diag_vs_early_relapse.cell_types.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)

# diagnostic vs. late relapse
comparators <- sort(unique(hl_all.combined$celltype.earlyrelapse))
# remove the "_earlyRelapse" and RLN cases
comparators <- comparators[ !grepl("_(earlyRelapse|RLN)", comparators)]

# perform the DE analysis
de.diag_vs_late_relapse.celltypes <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.diag_vs_late_relapse.celltypes.df <- merge_de_results(de.diag_vs_late_relapse.celltypes)
write.table(de.diag_vs_late_relapse.celltypes.df, 
            paste(work.dir, "de.diag_vs_late_relapse.cell_types.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)

## ------ cluster level -------

# define comparators
hl_all.combined$clusters.earlyrelapse <- paste(hl_all.combined$seurat_clusters, hl_all.combined$earlyRelapse, sep = "_")
Idents(hl_all.combined) <- hl_all.combined$clusters.earlyrelapse 
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$clusters.earlyrelapse))

# diagnostic vs. early relapse
comparators <- sort(unique(hl_all.combined$clusters.earlyrelapse))
# remove the "_lateRelapse" and RLN cases
comparators <- comparators[ !grepl("_(lateRelapse|RLN)", comparators)]

# perform the DE analysis
de.diag_vs_early_relapse.clusters <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.diag_vs_early_relapse.clusters.df <- merge_de_results(de.diag_vs_early_relapse.clusters)
write.table(de.diag_vs_early_relapse.clusters.df, 
            paste(work.dir, "de.diag_vs_early_relapse.clusters.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)

# diagnostic vs. late relapse
comparators <- sort(unique(hl_all.combined$clusters.earlyrelapse))
# remove the "_earlyRelapse" and RLN cases
comparators <- comparators[ !grepl("_(earlyRelapse|RLN)", comparators)]

# perform the DE analysis
de.diag_vs_late_relapse.clusters <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.diag_vs_late_relapse.clusters.df <- merge_de_results(de.diag_vs_late_relapse.clusters)
write.table(de.diag_vs_late_relapse.clusters.df, 
            paste(work.dir, "de.diag_vs_late_relapse.clusters.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)


## #########################################################
## Cluster-level DE analysis: refractory vs. relapse
## #########################################################

## ------ cell type level -------

# define comparators
Idents(hl_all.combined) <- hl_all.combined$celltype.refractory
comparators <- rev(sort(unique(hl_all.combined$celltype.refractory)))

# remove the "_diagnostic" and RLN cases
comparators <- comparators[ !grepl("_(diagnostic|RLN)", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$celltype.refractory))

# perform the DE analysis
de.refractory_vs_relapse.celltypes <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.refractory_vs_relapse.celltypes.df <- merge_de_results(de.refractory_vs_relapse.celltypes)
write.table(de.refractory_vs_relapse.celltypes.df, 
            paste(work.dir, "de.refractory_vs_relapse.cell_types.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)

## ------ cluster level -------

# define comparators
Idents(hl_all.combined) <- hl_all.combined$cluster.refractory
comparators <- rev(sort(unique(hl_all.combined$cluster.refractory)))

# remove the "_diagnostic" and RLN cases
comparators <- comparators[ !grepl("_(diagnostic|RLN)", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$cluster.refractory))

# perform the DE analysis
de.refractory_vs_relapse.clusters <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.refractory_vs_relapse.clusters.df <- merge_de_results(de.refractory_vs_relapse.clusters)
write.table(de.refractory_vs_relapse.clusters.df, 
            paste(work.dir, "de.refractory_vs_relapse.clusters.txt", sep = ""), 
            sep = "\t", quote = FALSE, row.names = FALSE)


## #########################################################
## Cluster-level DE analysis: early relapse vs. late relapse
## #########################################################

## ------ cell type level -------

# define comparators
Idents(hl_all.combined) <- hl_all.combined$celltype.earlyrelapse 
comparators <- sort(unique(hl_all.combined$celltype.earlyrelapse))

# remove the "_diagnostic" and RLN cases
comparators <- comparators[ !grepl("_(diagnostic|RLN)", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$celltype.earlyrelapse))

# perform the DE analysis
de.early_vs_late_relapse.celltypes <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.early_vs_late_relapse.celltypes.df <- merge_de_results(de.early_vs_late_relapse.celltypes)
write.table(de.early_vs_late_relapse.celltypes.df, 
            paste(work.dir, "de.early_vs_late_relapse.cell_types.txt", sep = ""),
            sep = "\t", quote = FALSE, row.names = FALSE)

## ------ cluster level -------

# define comparators
Idents(hl_all.combined) <- hl_all.combined$clusters.earlyrelapse 
comparators <- sort(unique(hl_all.combined$clusters.earlyrelapse))

# remove the "_diagnostic" and RLN cases
comparators <- comparators[ !grepl("_(diagnostic|RLN)", comparators)]
# store number of cells for each comparator
cell_counts.df <- as.data.frame(table(hl_all.combined$clusters.earlyrelapse))

# perform the DE analysis
de.early_vs_late_relapse.clusters <- get_de_genes(comparators, de.test = "MAST", cell_counts.df)
de.early_vs_late_relapse.clusters.df <- merge_de_results(de.early_vs_late_relapse.clusters)
write.table(de.early_vs_late_relapse.clusters.df, 
            paste(work.dir, "de.early_vs_late_relapse.clusters.txt", sep = ""),
            sep = "\t", quote = FALSE, row.names = FALSE)
```

## Write out subsetted expression data (maybe useful for iTALK?)

```{r}
library(dplyr)

# extract expression profiles for macrophage and "T cell" annotated cell barcodes
tcell_cluster_index <- 1
tcell_clusters <- c("NaÃ¯ve T cells", "CTL", "T regs", "Helper T cells")
macrophage_cluster <- c("Macrophages")
# clusters.keep <- c(macrophage_cluster, tcell_clusters[tcell_cluster_index])
# get normalized expression values -- rownames are genes, colnames are cell barcodes
norm.expr <- GetAssayData(object = hl.both.combined[["RNA"]], slot = "data")
# get barcodes for cell type
barcodes.celltype <- as.data.frame(hl.both.combined$celltype)
# get barcodes for diagnostic / relapse
barcodes.diag_vs_rel <- as.data.frame(hl.both.combined$case.type)
# get barcodes for diagnostic / early relapse
barcodes.diag_vs_early_rel <- as.data.frame(hl.both.combined$earlyRelapse)

# create filtered expression matrix containing only macrophage and t cell clusters
clusters.keep <- c(macrophage_cluster, tcell_clusters)
barcodes.celltype.cluster_filtered <- filter(barcodes.celltype, barcodes.celltype$`hl.both.combined$celltype` %in% clusters.keep)
expr.cluster_filtered <- norm.expr[, colnames(norm.expr) %in% rownames(barcodes.celltype.cluster_filtered)]

## IMPORTANT: before appending columns, will likely need to output as matrix (since current object is dgCMatrix)
write_dgCMatrix_csv(expr.cluster_filtered, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/hl_combined.norm_expr.macrophage_t_cell_clusters.txt", col1_name = "gene", chunk_size = 500)
expr.cluster_filtered.df <- t(read.table("~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/hl_combined.norm_expr.macrophage_t_cell_clusters.txt", sep = ",", header = TRUE, stringsAsFactors = FALSE))
cols <- expr.cluster_filtered.df[1, ]
colnames(expr.cluster_filtered.df) <- cols
expr.cluster_filtered.df <- expr.cluster_filtered.df[-1, ]
# rename row names to replace "." with "-" so they can be compared to barcode annotations
rownames(expr.cluster_filtered.df) <- gsub('\\.', '-', rownames(expr.cluster_filtered.df))

# include "cell_type" column
expr.cluster_filtered.df.anno <- merge(expr.cluster_filtered.df, barcodes.celltype, by="row.names")
rownames(expr.cluster_filtered.df.anno) <- expr.cluster_filtered.df.anno$Row.names

# include "compare_group" column indicating diagnostic / relapse OR diagnostic / early relapse
expr.cluster_filtered.df.anno <- merge(expr.cluster_filtered.df.anno, barcodes.diag_vs_rel, by="row.names")
rownames(expr.cluster_filtered.df.anno) <- expr.cluster_filtered.df.anno$Row.names

# write out to ~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/expr/
expr.cluster_filtered.df.anno <- merge(expr.cluster_filtered.df.anno, barcodes.diag_vs_early_rel, by="row.names")
rownames(expr.cluster_filtered.df.anno) <- expr.cluster_filtered.df.anno$Row.names

# clean up the matrix
expr.cluster_filtered.df.anno$Row.names <- NULL
colnames(expr.cluster_filtered.df.anno)[colnames(expr.cluster_filtered.df.anno) == 'Row.names'] <- 'barcode'
colnames(expr.cluster_filtered.df.anno)[colnames(expr.cluster_filtered.df.anno) == 'hl.both.combined$celltype'] <- 'cell_type'
colnames(expr.cluster_filtered.df.anno)[colnames(expr.cluster_filtered.df.anno) == 'hl.both.combined$case.type'] <- 'diag_rel'
colnames(expr.cluster_filtered.df.anno)[colnames(expr.cluster_filtered.df.anno) == 'hl.both.combined$earlyRelapse'] <- 'diag_early_rel'

# write out for input into iTALK
write.table(expr.cluster_filtered.df.anno, "~/Documents/sc_HL_relapse/seurat/output/diagnostic_plus_relapse/hl_combined.norm_expr.macrophage_t_cell_clusters.annotated.txt", sep = "\t")
```

## Visualizing differential expression

```{r}
# violin plot split by dataset (primary vs. relapse) per gene showing expression across clusters ("celltype")
# hl.both.combined[["celltype"]] <- Idents(hl.both.combined)
Idents(hl_all.combined) <- hl_all.combined$seurat_clusters

# B and T cells
plots <- VlnPlot(hl_all.combined, features = c("CD3D", "MS4A1"), 
                 split.by = "case.type", group.by = "celltype", pt.size = 0, combine = FALSE)
CombinePlots(plots = plots, ncol = 1)

# CTLA4 and LAG3
plots <- VlnPlot(hl_all.combined, features = c("CTLA4", "LAG3"), 
                 split.by = "case.type", group.by = "celltype", pt.size = 0, combine = FALSE)
CombinePlots(plots = plots, ncol = 1)

# genes of interest: "JUND", PD-L1 --> CD274 and  only found in RNA; 
# CXCL12 (no alias? filtered out early on?)
# CD123 --> IL3RA, CD30 --> TNFRSF8 
genes <- c("PDCD1", "TIGIT", "LAG3", "HAVCR2", "CD274", "CD163", "IL3RA", "GZMA", "GZMB", "GZMK", "PRF1", "IL2RA", "CCR4", "FOXP3", "CXCL13", "CXCR5", "CXCR4", "CXCR3", "CXCL9", "CXCL10", "CXCL12", "IFNG", "IL2", "IL4", "TNFRSF8", "JUND", "HLA-DRB5", "MKI67", "TOX", "CD27", "CD44", "CXCL11", "CD80", "CD86", "CD40", "CCL22", "CD28")
genes.low <- c("CD274", "CD163", "CCR4", "FOXP3", "IL2", "IL4", "TNFRSF8", "CXCL11", "CD80", "CCL22")

genes.diag <- c("CXCL9", "CXCL10", "IFNG")
genes.rel <- c("JUND")
genes.CXC <- c("CXCR3", "CXCR4", "CXCR5", "CXCR9", "CXCL10", "CXCL11")
gene.temp <- c("FOXP3")

# plot genes with low expression
p <- VlnPlot(hl_all.combined, features = gene.temp)

# plot GOI per cluster, split by case type (diagnostic, relapse, RLN)
p.by_case <- VlnPlot(hl_all.combined, features = genes, 
        split.by = "case.type", group.by = "seurat_clusters", 
        pt.size = 0, ncol = 1, log = FALSE, sort = "increasing", 
        assay = "RNA", slot = "data") + xlab("")

p.by_case.low <- VlnPlot(hl_all.combined, features = genes.low,
                         split.by = "case.type", group.by = "seurat_clusters",
                         pt.size = 0.1, ncol = 1, sort = "increasing", log = FALSE, 
                         assay = "RNA", slot = "data") + xlab("")

p.by_early_rel <- VlnPlot(hl_all.combined, features = genes, 
        split.by = "earlyRelapse", group.by = "seurat_clusters", 
        pt.size = 0, ncol = 1, log = FALSE, sort = "increasing", 
        assay = "RNA", slot = "data") + xlab("")

p.by_early_rel.low <- VlnPlot(hl_all.combined, features = genes.low,
                         split.by = "earlyRelapse", group.by = "seurat_clusters",
                         log = FALSE, pt.size = 0.1, ncol = 1, sort = "increasing",
                         assay = "RNA", slot = "data") + xlab("")

# JUND in all clusters (split by diagnostic, relapse, RLN)
p.JUND.clusters <- VlnPlot(hl_all.combined, features = c("JUND"),
                         split.by = "case.type", group.by = "seurat_clusters",
                         log = FALSE, pt.size = 0, ncol = 1, 
                         assay = "RNA", slot = "data") + xlab("")

## Is JUND consistently upregulated in all relapse samples?
p.JUND.samples <- VlnPlot(hl_all.combined, features = c("JUND"),
                         split.by = "case.type", group.by = "orig.ident",
                         log = FALSE, pt.size = 0, ncol = 1, 
                         assay = "RNA", slot = "data") + xlab("")

# can add boxplot on top: VlnPlot(object = seurat_mat_filtered, features = markers_filt[[i]], pt.size = 0)+geom_boxplot(width=0.1,fill="white")

p.temp <- VlnPlot(hl_all.combined, features = c("JUND"),
                         split.by = "earlyRelapse", group.by = "orig.ident",
                         log = FALSE, pt.size = 0, ncol = 1, 
                         assay = "RNA", slot = "data") + xlab("")
```

## Violin plots with statistical bars

```{r}
library(ggpubr)

vp_case1 <- function(gene_signature, file_name, test_sign, y_max, curr_idents, width, height){
  plot_case1 <- function(signature){
    VlnPlot(hl_all.combined, features = signature,
            pt.size = 0,
            group.by = "case.type", idents = curr_idents,
            y.max = y_max, # add the y-axis maximum value - otherwise p-value hidden
            assay = "RNA", slot = "data"
    ) + stat_compare_means(comparisons = test_sign, label = "p.format")
  }
  purrr::map(gene_signature, plot_case1) %>% cowplot::plot_grid(plotlist = .)
  file_name <- paste(file_name, ".png", sep = "")
  ggsave(file_name, width = width, height = height)
}

# ######################
# diagnostic VS relapse
# ######################

# 1. pDC cluster: GZB, CD163, CD123 (-> IL3RA), HLA-DRB5
vp_case1(gene_signature = c("GZMB", "CD163", "IL3RA", "HLA-DRB5"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/pDC-diag_vs_rel", 
         test_sign = list(c("diagnostic", "relapse")), 
         y_max = 7, 
         c("Plasmacytoid DCs"), 10, 10)

# 2. Macrophage cluster: CD163, PD-L1 (-> CD274), CD123 (-> IL3RA), CD80, CD86, CXCL9, CXCL10, CXCL11, CD40, CCL22, TIM3 (-> HAVCR2), HLA-DRB5
vp_case1(gene_signature = c("CD163", "CD274", "IL3RA", "CD80", "CD86", "CXCL9", "CXCL10", "CXCL11", "CD40", "CCL22", "HAVCR2", "HLA-DRB5"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/macrophage-diag_vs_rel", 
         test_sign = list(c("diagnostic", "relapse")), 
         y_max = 6, 
         c("Macrophage"), width=20, height=10)

# 3. Treg cluster (cluster 4): LAG3, FOXP3, TIM3 (-> HAVCR2), CXCR3, PD-1 (-> PDCD1), Ki67 (-> MKI67), CD30 (-> TNFRSF8), TOX, CXCR4

# 4. Treg (cell type): LAG3, FOXP3, TIM3 (-> HAVCR2), CXCR3, PD-1 (-> PDCD1), Ki67 (-> MKI67), CD30 (-> TNFRSF8), TOX, CXCR4
vp_case1(gene_signature = c("LAG3", "FOXP3", "HAVCR2", "CXCR3", "PDCD1", "MKI67", "TNFRSF8", "TOX", "CXCR4"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/Treg-diag_vs_rel", 
         test_sign = list(c("diagnostic", "relapse")), 
         y_max = 6, 
         c("Treg"), width=20, height=10)

# 5. Helper T cell (cluster 1): LAG3, CXCR5, TIM3 (HAVCR2), CXCR3, PD-1 (PDCD1), Ki67 (MKI67), CD30 (TNFRSF8), CXCL13, TOX, CD28, IL-4 (IL4), CXCR4


# 6. Helper T cell (cell type): LAG3, CXCR5, TIM3 (HAVCR2), CXCR3, PD-1 (PDCD1), Ki67 (MKI67), CD30 (TNFRSF8), CXCL13, TOX, CD28, IL-4 (IL4), CXCR4
vp_case1(gene_signature = c("LAG3", "CXCR5", "HAVCR2", "CXCR3", "PDCD1", "MKI67", "TNFRSF8", "CXCL13", "TOX", "CD28", "IL4", "CXCR4"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/HelperT-diag_vs_rel", 
         test_sign = list(c("diagnostic", "relapse")), 
         y_max = 7.5, 
         c("Helper T cells"), width=20, height=10)

# 7. CTL (cluster 8): GZMA, GZMB, GZMK, PRF1, CXCR3, IFNG, TIM3 (HAVCR2), LAG3, PD1 (PDCD1), CXCR4


# 8. CTL (cell type): GZMA, GZMB, GZMK, PRF1, CXCR3, IFNG, TIM3 (HAVCR2), LAG3, PD1 (PDCD1), CXCR4
vp_case1(gene_signature = c("GZMA", "GZMB", "GZMK", "PRF1", "CXCR3", "IFNG", "HAVCR2", "LAG3", "PDCD1", "CXCR4"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/CTL-diag_vs_rel", 
         test_sign = list(c("diagnostic", "relapse")), 
         y_max = 7, 
         c("CTL"), width=20, height=10)

# 9. NK cells: GZMA, GZMB, GZMK, PRF1, CXCR3, IFNG
vp_case1(gene_signature = c("GZMA", "GZMB", "GZMK", "PRF1", "CXCR3", "IFNG"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/NK-diag_vs_rel", 
         test_sign = list(c("diagnostic", "relapse")), 
         y_max = 7, 
         c("NK cells"), width=20, height=10)

# ############################
# diagnostic VS early relapse
# ############################

vp_case1(gene_signature = c("LAG3", "FOXP3", "HAVCR2", "CXCR3", "PDCD1", "MKI67", "TNFRSF8", "TOX", "CXCR4"),
         file_name = "~/Documents/sc_HL_relapse/seurat/figures/LATEST/violin_with_p_vals/Treg-diag_vs_early_rel", 
         test_sign = list(c("diagnostic", "earlyRelapse")), 
         y_max = 6, 
         c("Treg"), width=20, height=10)

```

```{r}
### Examples using cowplot ###
title <- ggdraw() + 
  draw_label("Miles per gallon decline with displacement and horsepower", fontface = 'bold', x = 0, hjust = 0) +
  theme(plot.margin = margin(0, 0, 0, 7))

plot_grid(title, plot_row, ncol = 1,
  # rel_heights values control vertical title margins
  rel_heights = c(0.1, 1)
)
```

## Volcano plots

```{r}
library(EnhancedVolcano)
library(ggplot2)
library(dplyr)
library(plyr)
library(tidyr) # separate

# define working directory:
work.dir <- "~/Documents/sc_HL_relapse/seurat/output/LATEST/DE_analysis/"

de.diag_vs_rel.celltypes.df <- read.table(paste(work.dir, "de.diag_vs_rel.cell_types.txt", sep = ""), 
                                         header = TRUE, sep = "\t")
de.diag_vs_refractory.celltypes.df <- read.table(paste(work.dir, "de.diag_vs_refractory.cell_types.txt", sep = ""), 
                                                header = TRUE, sep = "\t")
de.diag_vs_early_relapse.celltypes.df <- read.table(paste(work.dir, "de.diag_vs_early_relapse.cell_types.txt", sep = ""), 
                                                   header = TRUE, sep = "\t")
de.diag_vs_late_relapse.celltypes.df <- read.table(paste(work.dir, "de.diag_vs_late_relapse.cell_types.txt", sep = ""),
                                                  header = TRUE, sep = "\t")
de.refractory_vs_relapse.celltypes.df <- read.table(paste(work.dir, "de.refractory_vs_relapse.cell_types.txt", sep = ""),
                                                   header = TRUE, sep = "\t")
de.early_vs_late_relapse.celltypes.df <- read.table(paste(work.dir, "de.early_vs_late_relapse.cell_types.txt", sep = ""),
                                                   header = TRUE, sep = "\t")

results <- list(de.diag_vs_rel.celltypes.df,
             de.diag_vs_refractory.celltypes.df,
             de.diag_vs_early_relapse.celltypes.df,
             de.diag_vs_late_relapse.celltypes.df,
             de.refractory_vs_relapse.celltypes.df,
             de.early_vs_late_relapse.celltypes.df)

i <- 6
titles <- c("diagnostic vs. relapse", "diagnostic vs. refractory", "diagnostic vs. early relapse",
            "diagnostic vs. late relapse", "refractory vs. relapse", "early vs. late relapse")
title <- titles[i]
test.df <- results[[i]]

#split "comparison" column by "_" and take the first split element to be cell_type
test.df <- separate(data = test.df, col = comparison, into = c("cell_type", "group1"), sep = "\\_")
test.df$group1 <- NULL

LOGFC_CUTOFF = 0.25
PVAL_CUTOFF = 0.05

keyvals.colour <- ifelse((abs(test.df$avg_logFC) > LOGFC_CUTOFF & test.df$p_val_adj < PVAL_CUTOFF), 'firebrick3',
      ifelse((abs(test.df$avg_logFC) > LOGFC_CUTOFF & test.df$p_val_adj >= PVAL_CUTOFF), 'forestgreen',
             ifelse((abs(test.df$avg_logFC) < LOGFC_CUTOFF & test.df$p_val_adj < PVAL_CUTOFF), 'dodgerblue3',
        'gray32')))
keyvals.colour[is.na(keyvals.colour)] <- 'black'
names(keyvals.colour)[keyvals.colour == 'firebrick3'] <- 'q-val < 0.05; |logFC| > 0.25'
names(keyvals.colour)[keyvals.colour == 'forestgreen'] <- '|logFC| > 0.25'
names(keyvals.colour)[keyvals.colour == 'dodgerblue3'] <- 'q-val < 0.05'
names(keyvals.colour)[keyvals.colour == 'gray32'] <- 'unsignificant'

## ------ Enhanced volcano plot ------

p.e <- EnhancedVolcano(test.df,
    lab = test.df$gene,
    title = title, subtitle = "",
    x = 'avg_logFC',
    y = 'p_val_adj', 
    ylim = c(-10, 300), 
    xlim = c(-2, 2.5),
    captionLabSize = 0,
    FCcutoff = LOGFC_CUTOFF,
    pCutoff = PVAL_CUTOFF,
    colCustom = keyvals.colour,
    legendPosition = 'bottom')

p.facets.grid <- p.e + facet_wrap(~ cell_type, ncol=5) +
  theme(strip.text.x = element_text(size = 10))


## ------ Custom Enhanced volcano plots ------

## diag vs. rel, diag vs. late relapse - LAG3, PD-1 (PDCD1), PD-L1 (CD274), GZB, CD80, CD86, TIM3 (HAVCR2)
# define different cell-types that will be shaded
GOI <- c('LAG3', 'PDCD1', 'CD274', 'GZMB', 'CD80', 'CD86', 'HAVCR2')
# create custom key-value pairs for different cell-types
# this can be achieved with nested ifelse statements
keyvals.shape <- ifelse(test.df$gene %in% GOI, 17, 1)
keyvals.shape[is.na(keyvals.shape)] <- 1
names(keyvals.shape)[keyvals.shape == 1] <- 'all other genes'
names(keyvals.shape)[keyvals.shape == 17] <- 'GOI'

p.e.0 <- EnhancedVolcano(test.df,
    lab = test.df$gene,
    title = title, subtitle = "",
    selectLab = GOI,
    colAlpha = 0.9,
    drawConnectors = TRUE,
    widthConnectors = 0.5,
    pointSize = c(ifelse(test.df$gene %in% GOI, 8, 1)),
    labSize = 5.0,
    captionLabSize = 0,
    x = 'avg_logFC',
    y = 'p_val_adj',
    FCcutoff = LOGFC_CUTOFF,
    pCutoff = PVAL_CUTOFF,
    shapeCustom = keyvals.shape,
    colCustom = keyvals.colour, 
    legendPosition = 'bottom') 

p.facets.grid <- p.e.0 + facet_wrap(~ cell_type, ncol=5) +
  theme(strip.text.x = element_text(size = 10))

# for combined volcano plot (all clusters plotted together in one plot), map shapes to cell types

# create custom key-value pairs for different cell-types
  # this can be achieved with nested ifelse statements
# keyvals.shape <-  ifelse(test.df$cell_type == 'B and T mix', 14,
#                   ifelse(test.df$cell_type == 'B cells', 0,
#                   ifelse(test.df$cell_type == 'B cells, macrophages', 15,
#                   ifelse(test.df$cell_type == 'CTL', 12,
#                   ifelse(test.df$cell_type == 'GCB', 3,
#                   ifelse(test.df$cell_type == 'Helper T cells', 17,       
#                   ifelse(test.df$cell_type == 'Macrophage', 10, 
#                   ifelse(test.df$cell_type == 'Naive T cells', 2,
#                   ifelse(test.df$cell_type == 'NK cells', 4,
#                   ifelse(test.df$cell_type == 'Treg', 18,
#                   ifelse(test.df$cell_type == 'Treg proliferation', 8, 
#                   19)))))))))))
# 
# keyvals.shape[is.na(keyvals.shape)] <- 19
# names(keyvals.shape)[keyvals.shape == 19] <- 'other'
# names(keyvals.shape)[keyvals.shape == 14] <- 'B and T mix'
# names(keyvals.shape)[keyvals.shape == 0] <- 'B cells'
# names(keyvals.shape)[keyvals.shape == 15] <- 'B cells, macrophages'
# names(keyvals.shape)[keyvals.shape == 12] <- 'CTL'
# names(keyvals.shape)[keyvals.shape == 3] <- 'GCB'
# names(keyvals.shape)[keyvals.shape == 17] <- 'Helper T cells'
# names(keyvals.shape)[keyvals.shape == 10] <- 'Macrophage'
# names(keyvals.shape)[keyvals.shape == 2] <- 'Naive T cells'
# names(keyvals.shape)[keyvals.shape == 4] <- 'NK cells'
# names(keyvals.shape)[keyvals.shape == 18] <- 'Treg'
# names(keyvals.shape)[keyvals.shape == 8] <- 'Treg proliferation'

## ------ Enhanced volcano plot ------

# previous p-value cutoff: 10e-32; excludes CXCL10 in macrophages
# p.e <- EnhancedVolcano(test.df,
#     lab = test.df$gene,
#     title = title, subtitle = "",
#     x = 'avg_logFC',
#     y = 'p_val_adj',
#     FCcutoff = LOGFC_CUTOFF,
#     pCutoff = PVAL_CUTOFF,
#     shapeCustom = keyvals.shape,
#     colCustom = keyvals.colour,
#     legendPosition = 'right')
# 
# p.facets.grid <- p.e + facet_wrap(~ cell_type, ncol=2) +
#   theme(strip.text.x = element_text(size = 10))

```


## EnhancedVolcano plots for clusters (no mappings)

```{r}
# define working directory:
work.dir <- "~/Documents/sc_HL_relapse/seurat/output/LATEST/DE_analysis/"

de.diag_vs_rel.clusters.df <- read.table(paste(work.dir, "de.diag_vs_rel.cell_types.txt", sep = ""), 
                                         header = TRUE, sep = "\t")
de.diag_vs_refractory.clusters.df <- read.table(paste(work.dir, "de.diag_vs_refractory.cell_types.txt", sep = ""), 
                                                header = TRUE, sep = "\t")
de.diag_vs_early_relapse.clusters.df <- read.table(paste(work.dir, "de.diag_vs_early_relapse.cell_types.txt", sep = ""), 
                                                   header = TRUE, sep = "\t")
de.diag_vs_late_relapse.clusters.df <- read.table(paste(work.dir, "de.diag_vs_late_relapse.cell_types.txt", sep = ""),
                                                  header = TRUE, sep = "\t")
de.refractory_vs_relapse.clusters.df <- read.table(paste(work.dir, "de.refractory_vs_relapse.cell_types.txt", sep = ""),
                                                   header = TRUE, sep = "\t")
de.early_vs_late_relapse.clusters.df <- read.table(paste(work.dir, "de.early_vs_late_relapse.cell_types.txt", sep = ""),
                                                   header = TRUE, sep = "\t")

results <- list(de.diag_vs_rel.clusters.df,
             de.diag_vs_refractory.clusters.df,
             de.diag_vs_early_relapse.clusters.df,
             de.diag_vs_late_relapse.clusters.df,
             de.refractory_vs_relapse.clusters.df,
             de.early_vs_late_relapse.clusters.df)

i <- 6
titles <- c("diagnostic vs. relapse", "diagnostic vs. refractory", "diagnostic vs. early relapse",
            "diagnostic vs. late relapse", "refractory vs. relapse", "early vs. late relapse")
title <- titles[i]
test.df <- results[[i]]

#split "comparison" column by "_" and take the first split element to be cell_type
test.df <- separate(data = test.df, col = comparison, into = c("cluster", "group1"), sep = "\\_")
test.df$group1 <- NULL

LOGFC_CUTOFF = 0.25
PVAL_CUTOFF = 0.05

keyvals.colour <- ifelse((abs(test.df$avg_logFC) > LOGFC_CUTOFF & test.df$p_val_adj < PVAL_CUTOFF), 'firebrick3',
      ifelse((abs(test.df$avg_logFC) > LOGFC_CUTOFF & test.df$p_val_adj >= PVAL_CUTOFF), 'forestgreen',
             ifelse((abs(test.df$avg_logFC) < LOGFC_CUTOFF & test.df$p_val_adj < PVAL_CUTOFF), 'dodgerblue3',
        'gray32')))
keyvals.colour[is.na(keyvals.colour)] <- 'black'
names(keyvals.colour)[keyvals.colour == 'firebrick3'] <- 'q-val < 0.05; |logFC| > 0.25'
names(keyvals.colour)[keyvals.colour == 'forestgreen'] <- '|logFC| > 0.25'
names(keyvals.colour)[keyvals.colour == 'dodgerblue3'] <- 'q-val < 0.05'
names(keyvals.colour)[keyvals.colour == 'gray32'] <- 'unsignificant'

## ------ Enhanced volcano plot ------

p.e <- EnhancedVolcano(test.df,
    lab = test.df$gene,
    title = title, subtitle = "",
    x = 'avg_logFC',
    y = 'p_val_adj', 
    ylim = c(-12, 300), 
    xlim = c(-2, 2.8),
    captionLabSize = 0,
    FCcutoff = LOGFC_CUTOFF,
    pCutoff = PVAL_CUTOFF,
    colCustom = keyvals.colour,
    legendPosition = 'bottom')

p.facets.grid <- p.e + facet_wrap(~ cluster, ncol=6) +
  theme(strip.text.x = element_text(size = 10))


## ------ Custom Enhanced volcano plots ------

## diag vs. rel, diag vs. late relapse - LAG3, PD-1 (PDCD1), PD-L1 (CD274), GZB, CD80, CD86, TIM3 (HAVCR2)
# define different cell-types that will be shaded
GOI <- c('LAG3', 'PDCD1', 'CD274', 'GZMB', 'CD80', 'CD86', 'HAVCR2')
# create custom key-value pairs for different cell-types
# this can be achieved with nested ifelse statements
keyvals.shape <- ifelse(test.df$gene %in% GOI, 17, 1)
keyvals.shape[is.na(keyvals.shape)] <- 1
names(keyvals.shape)[keyvals.shape == 1] <- 'all other genes'
names(keyvals.shape)[keyvals.shape == 17] <- 'GOI'


p.e.0 <- EnhancedVolcano(test.df,
    lab = test.df$gene,
    title = title, subtitle = "",
    selectLab = GOI,
    colAlpha = 0.9,
    captionLabSize = 0,
    drawConnectors = TRUE,
    widthConnectors = 0.5,
    pointSize = c(ifelse(test.df$gene %in% GOI, 8, 1)),
    labSize = 5.0,
    x = 'avg_logFC',
    y = 'p_val_adj',
    FCcutoff = LOGFC_CUTOFF,
    pCutoff = PVAL_CUTOFF,
    shapeCustom = keyvals.shape,
    colCustom = keyvals.colour, 
    legendPosition = 'bottom') 

p.facets.grid <- p.e.0 + facet_wrap(~ cluster, ncol=6) +
  theme(strip.text.x = element_text(size = 10))

## diag vs. rel, diag vs. late relapse - CXCL10

p.e.1 <- EnhancedVolcano(test.df,
    lab = test.df$gene,
    title = title, subtitle = "",
    selectLab = c('CXCL10'),
    drawConnectors = TRUE,
    xlim = c(-2,2.5),
    labSize = 6.0,
    x = 'avg_logFC',
    y = 'p_val_adj',
    FCcutoff = LOGFC_CUTOFF,
    pCutoff = PVAL_CUTOFF,
    colCustom = keyvals.colour,
    legendPosition = 'bottom')

## diag vs. refractory - CXCL13, CXCR4, CXCR6

p.e.2 <- EnhancedVolcano(test.df,
    lab = test.df$gene,
    title = title, subtitle = "",
    selectLab = c('CXCL13', 'CXCR4', 'CXCR6'),
    drawConnectors = TRUE,
    #xlim = c(-2,2.5),
    labSize = 6.0,
    x = 'avg_logFC',
    y = 'p_val_adj',
    FCcutoff = LOGFC_CUTOFF,
    pCutoff = PVAL_CUTOFF,
    colCustom = keyvals.colour,
    legendPosition = 'bottom')

```

```{r}

## --- manual volcano plot using ggplot2 ---

y_axis_label <- expression(paste("-",Log[10], "(q-value)"))
x_axis_label <- "Avg LogFC"

# add annotation to de results for colouring on volcano plots - "group"
#   min columns: gene, logFC, adjusted p-value, group
#   groups: not sig + low FC, not sig + high FC, sig + low logFC, sig + high FC
# label top 20 genes that are sig + high FC

# x-axis: fold change (e.g. from -3 to +3)
# y-axis: -log10(adj p-value)

# filter out genes that have a low p-value and low fc
# test.df <- filter(test.df, !(test.df$p_val_adj < 1e-100 & abs(test.df$avg_logFC) < 0.1))

# assign groups for plot colors
test.df <- test.df %>% mutate(
  group = case_when(
    p_val_adj < 0.05 & abs(avg_logFC) < 0.25 ~ "Significant, low logFC",
    p_val_adj < 0.05 & abs(avg_logFC) >= 0.25 ~ "Significant, high logFC",
    p_val_adj >= 0.05 & abs(avg_logFC) < 0.25 ~ "Not significant, low logFC",
    p_val_adj >= 0.05 & abs(avg_logFC) >= 0.25 ~ "Not significant, high logFC"
  )
)

# transform the p-values so they can be used for the volcano plot
test.df$pValTrans <- -log10(test.df$p_val_adj)

# plot genes of interest (to be labelled)
p <- ggplot(test.df, aes(x=avg_logFC, y=pValTrans, color=group)) + 
  geom_point(size = 2) +
  labs(y=y_axis_label, x=x_axis_label, color="p < 0.05") +
  scale_colour_manual(name="", 
                      values = c("Not significant, high logFC"="darkgreen", 
                                 "Not significant, low logFC"="darkgrey",
                                 "Significant, low logFC"="royalblue", 
                                 "Significant, high logFC"="red"),
                      labels = c("Not significant, high logFC", 
                                 "Not significant, low logFC", 
                                 "Significant, low logFC",
                                 "Significant, high logFC")) + 
  theme(legend.position="bottom") 
  # xlim(-2.5,2.5)

p.facets <- p + facet_grid(rows = vars(comparison), scales = "free") + theme_gray(base_size = 14) 


  # geom_hline(yintercept = -log10(0.05), linetype="dashed") +
  # geom_vline(xintercept = -0.4849312,  linetype="dashed", color = "green") +
  # geom_vline(xintercept = 0.5176877,  linetype="dashed", color = "green") +
  # geom_vline(xintercept = -0.5264313,  linetype="dashed", color = "blue") +
  # geom_vline(xintercept = 0.5604279,  linetype="dashed", color = "blue") +

```


## Special request from Richard Corbett - what are the cell types for each barcode?

```{r}
library(reshape2)

# Retrieve cluster assignments (identities) for cell barcodes in cHL-LSARP-09, 10, 11, 12, 19, 20, 25, and 26
cluster.mappings <- read.table("~/Documents/sc_HL_relapse/data/relapse/hl_relapse-cluster_annotations.txt", sep = "\t", header = TRUE)
gsc.data <- read.table("~/Documents/sc_HL_relapse/data/relapse/library_info-relapse-cohort.txt", sep = "\t", header = TRUE, fill = TRUE)

# merge dataframes to combine barcode, cluster id, and cell type annotation
colnames(clusters.df) <- c("barcode_sample", "cluster")
clusters.annotated <- merge(clusters.df, cluster.mappings, by = "cluster")

# extract sample id so we can more easily filter for samples of interest
clusters.annotated <- cbind(clusters.annotated, colsplit(clusters.annotated$barcode_sample, pattern = "_", names = c("barcode", "sample_id")))

# save for future use
write.table(clusters.annotated, "~/Documents/sc_HL_relapse/seurat/output/relapse-all_25_merged/hl_relapse-merged.sct-clusters.annotated.txt", sep = "\t", row.names = FALSE, quote = FALSE)

# extract the essential columns
keeps <- c("sample_id", "barcode", "cluster", "cell_type")
clusters.annotated <- clusters.annotated[ , keeps ]

# now extract samples of interest
samples.keep <- c("cHL-LSARP-09", "cHL-LSARP-10", "cHL-LSARP-11", "cHL-LSARP-12",
                  "cHL-LSARP-19", "cHL-LSARP-20", "cHL-LSARP-25", "cHL-LSARP-26")
clusters.annotated.RC <- filter(clusters.annotated, clusters.annotated$sample_id %in% samples.keep)

# sort by sample id then cluster
clusters.annotated.RC <- clusters.annotated.RC[order(clusters.annotated.RC$sample_id, clusters.annotated.RC$cluster), ]

# add library information
gsc.library <- gsc.data[ c("Sample", "Library", "GSCPool") ]
colnames(gsc.library) <- c("sample_id", "gsc_library", "gsc_pool")
clusters.annotated.RC <- merge(clusters.annotated.RC, gsc.library, by = "sample_id")
keeps <- c("sample_id", "gsc_library", "gsc_pool", "barcode", "cluster", "cell_type")
clusters.annotated.RC <- clusters.annotated.RC[ , keeps]

# write to output
write.table(clusters.annotated.RC, "~/Documents/sc_HL_relapse/for_Richard/lsarp_hl_relapse-annotations-RC_samples.txt", sep = "\t", row.names = FALSE, quote = FALSE)
```

