---
title: "Seurat_analysis"
author: "Stacy Hung"
date: "02/12/2019"
output: html_document
---

This script goes through the Guided Clustering Tutorial, which analyzes the dataset of Peripheral Blood Mononuclear (PBMC) [from 10X Genomics] - this dataset has 2,700 single cells sequenced on the Illumina NextSeq 500.

See website for more details: https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html


```{r}
library(dplyr)
library(Seurat)
```

1. Read in the PBMC dataset

```{r}
# the Read10X function reads output of cellranger pipeline from 10X and returns a UMI count matrix - values in this matrix represent number of molecules for each feature (gene/row) that are detected in each cell (column)
pbmc.data <- Read10X("~/Documents/temp/seurat_tutorial/filtered_gene_bc_matrices/hg19/")

# for this dataset, the matrix is 32738 (genes) by 2700 (single cells)
```


2. Create a Seurat object - to serve as a container for both data (e.g. count matrix) and analyses (e.g. PCA, clustering, etc.) for a single-cell dataset.  See https://github.com/satijalab/seurat/wiki for technical details on the Seurat object.

```{r}
# Initialize the Seurat object with the raw (non-normalized data)
pbmc <- CreateSeuratObject(counts = pbmc.data, 
                           project = "pbmc3k", 
                           min.cells = 3, # feature cutoff - i.e. include genes detected in at least 3 genes
                           min.features = 200 # cell cutoff - i.e. include cells detected in at least 200 transcripts
                           )

pbmc

# after setting the cell and feature cutoffs, the count matrix is now 13714 (genes) X 2700 single cells
```

What does data in a count matrix look like?

```{r}
# examine a few genes in the first thirty cells
pbmc.data[c("CD3D", "TCL1A", "MS4A1"), 1:30]

# notice the "." values - these represent 0's (no molecules detected); as most values in a scRNAseq matrix are 0, Seurate uses a sparse-matrix representation whenever possible (significant memory + speed savings) for 10X data.
```

Standard pre-processing workflow - i.e. steps for:
  (i) selection and filtration of cells based on QC metrics
  (ii) data normalization and scaling, and 
  (iii) detection of highly variable features

3a. Pre-processing workflow - QC and selecting cells for further analysis

Metrics commonly used by the community:
- Number of unique genes detected in each cell
    e.g. low-quality cells / empty droplets will often have very few genes
    e.g. cell doublets / multiplets may exhibit an aberrantly high gene count
- Total number of molecules detected in each cell (strongly correlated with unique genes)
- Percent of reads that map to mitochondrial genome
    e.g. low-quality / dying cells often exhibit extensive mitochondrial contamination
    Note: Calculate mitochondrial QC metrics with `PercentageFeatureSet` function (calculates % counts originating from a set of features) --> use set of genes starting with "MT-" as set of mitochondrial genes

```{r}
# Get % reads that map to mitochondrial genome
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

# Get QC metrics stored in Seurat (automatically calculated when Seurat object is created)
# to get an idea, just show QC metrics for the first 5 cells:
head(pbmc@meta.data, 5)
```
Visualize QC metrics, and use these to filter cells:
* Filter cells that have unique feature counts > 2,500 | < 200
* Filter cells that have > 5% mitochondrial counts

```{r}
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```
```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA",  feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))

pbmc <- subset (pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
# this causes the count matrix to be 13714 genes (unchanged) X 2638 single cells (slightly reduced)
```

3b. Normalizing the data

Application of global-scaling normalization method `LogNormalize`, which normalizes feature expression measurements for each cell by total expression, multiplies this by a scale factor (10,000 by default) adn log-transforms the result.

```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
# normalized values in pbmc[["RNA"]]@data
```

3c. Identification of highly variable features (feature selection)

Identify features that have high cell-to-cell variation in the dataset (e.g. highly expressed in some cells, lowly expressed in others) --> focusing on these genes in downstream analysis helps to highlight biological signal. This is implemented in the function `FindVariableFeatures`, which returns 2000 features by default --> used in downstream analysis (e.g. PCA)

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# get the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))
```

3d. Application of a linear transformation ("scaling") (standard pre-processing step prior to dimensionality reduction techniques like PCA) -- `ScaleData` function:
- Shifts the expression of each gene, so that mean expression across cell is 0
- Scales the expression of each gene, so that variances across cells is 1
  - This step gives equal weight in downstream analysis, so highly expressed genes do not dominate
- Result is stored in pbmc[["RNA"]]@scale.data

```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

4. Perform linear dimensional reduction

```{r}
# perform PCA on the scaled data.  Only previously defined variable features are used as input, but can be re-defined using the features argument to choose a different subset
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))

# Examine and visualize PCA results in a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)

# plot loadings for specified PCA components (in this case, from 1 to 2)
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
```

`DimHeatmap` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide *which PCs* to include for further downstream analysis. Both cells and features are ordered according to their PCA scores.  Setting `cells` to a number plots the "extreme" cells on both ends of the spectrum, which dramatically speeds plotting for large dataset (e.g. 'cells = 500' --> plots 500 cells??)

Recommendations - err on the higher side for number of PCs to include; re-running analysis with a different number often will not change the results dramatically.

```{r}

# Dimension reduction plot - graph of output from dimension reduction on a 2D scatter
DimPlot(pbmc, reduction = "pca")

# Plot PCA dimension 1
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

# Plot the first 15 PCA dimensions
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

Determine the "dimensionality" of the dataset.

Seurat clusters cells based on PCA scores, with each PC essentially representing a "metafeature" that combines information across a correlated feature set.  The top principal components thus represent a robust compression of the dataset.... but how many components should we include ... 10? 20? 100?

`JacksonStrawPlot` function provides visualization tool for comparing distribution of p-values for each PC wiht a uniform distribution. "Significant" PCs will show a strong enrichment of features with low p-values.

```{r}
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
JackStrawPlot(pbmc, dims = 1:15)
```

Alternative heuristic method: "Elbow plot" - ranking of principle components based on % of variance explained by each one --> `ElbowPlot` function.

```{r}
ElbowPlot(pbmc) # in this example, see an elbow around 9-10, suggesting majority of true signal is in first 10 PCs
```

5. Cluster the cells

Seurat applies an dramatically improved graph-based clustering approach, driven by the distance metric (based on previously identified PCs) - inspired by SNN-Cliq and PhenoGraph.  Briefly, these methods embed cells in a graph structure (e.g. K-nearest neighbour graph) with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'. 

Overview of method:
1. Construct KNN graph based on euclidean distance in PCA space (similar to Phenograph)
2. Refine edge weights between any two cells based on shared overlap in local neighbourhoods (Jaccard similarity) --> `FindNeighbours` function (input = previously defined dimensionality of the dataset - i.e. first 10 PCs)
3. To cluster cells, apply modularity optimization techniques (e.g. Louvain algorithm - default, or SLM) to iteratively group cells together, with goal of optimizing the standard modularity function --> `FindClusters` (parameter for "granularity" of downstream clustering where increased values --> increased number of clusters; 0.4-1.2 generally returns good results for single-cell datasets ~3000 cells).
4. Obtain clusters with `Idents` function.

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

```{r}
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
```

Run non-linear dimensional reduction (UMAP/tSNE).

Seurat has a few non-linear dimensional reduction techniques (e.g. tSNE, UMAP) to visualize and explore these datasets.  The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space.  Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots.  

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:10)

DimPlot(pbmc, reduction = "umap")
#DimPlot(pbmc, reduction = "pca")

saveRDS(pbmc, "~/Documents/temp/seurat_tutorial/pbmc_tutorial.rds") # save as R file object (readRDS to read back in)
```

Finding differentially expressed features (cluster biomarkers).

Seurat can help to find markers that define clusters via differential expression.  By default, positive and negative markers of a single cluster (specified in `ident.l`) are identified compared to all other cells. `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other or against all cells.

The `min.pct` argument requires a feature to be detected at a minimum percentage in either of the two groups of cells and `thresh.test` arguement requires a feature to be differentially expression (on average) by some amount between the two groups (if both are set to 0 --> require much more time since will be testing a large number of features that are unlikely to be discriminatory); can also set `max.cells.per.iden` to speed up computations --> downsamples each identity class to have no more cells than specified threshold.

```{r}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, 
                                min.pct = 0.25, # default 0.1 (test genes detected in min.pct of cells, for either pop)
                                )
head(cluster1.markers, n=5)

```

```{r}
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0,3), min.pct = 0.25)
head(cluster5.markers, n=5)
```

```{r}
# find markers for every cluster compare to all remaining cells and report only positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

Seurat has several tests for differential expression that can be set with the test.use parameter - e.g. ROC test returns "classification power" for any individual marker (where 0 = random... 1 = perfect)

```{r}
cluster1.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)

# see https://satijalab.org/seurat/v3.0/de_vignette.html for other tests that can be used for differential expression.

# if you look at the head and tail of these results, you will notice that the results are sorted by decreasing power (i.e. top gene has the highest "classification power" - closest to 1)
```

Seurat provides several tools for visualizing marker expression - most commonly used visualizations:
(i) `VlnPlot` (shows expression probability distributions across clusters)
(ii) `FeaturePlot` (visualizes feature expression on a tSNE / PCA plot)

Other useful visualization methods: `RidgePlot`, `CellScatter`, `DotPlot`

```{r}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```

# can plot raw counts too

```{r}
VlnPlot(pbmc, features = c("NkG7", "PF4"), slot = "counts", log = TRUE)
```

The FeaturePlot:

```{r}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "LYZ", "PPBP", "CD8A"))
FeaturePlot(pbmc, features = c("MS4A1", "CD3E", "CD14", "LYZ", "PPBP"))
```

`DoHeatmap` generates an expression heatmap for given cells and features. In this case, we are plotting top 20 markers (or all markers if there are less than 20) for each cluster.

```{r}
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n=10, wt=avg_logFC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

Assigning cell type identity to clusters.

For this particular dataset, we are able to use canonical markers to easily match the unbiased clustering to known cell types.

```{r}
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids) # function for retrieving, setting, and manipulating an object's identity classes
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5)
```


```{r}
saveRDS(pbmc, file = "~/Documents/temp/seurat_tutorial/pbmc3k_final.rds")
```

